### PR TITLE
Fixed bugs and private/protected methods in CBOT

### LINKED ISSUE #207: Use of 'object::' notation in CBot for other classes
It is possible to abuse 'object::' notation used for accessing bot parameters in functions. You can do so by creating new class and using this class's name instead of 'object'. This will give you access to class's attributes like they were declared variables. These attributes are actually bot's parameters like category, position, orientation etc., but with different types and names. Attempt to modify them may result in crashing the game.

```
extern void object::CrashGame()
{
    crash();
}

public class Crasher
{
    int x;
    float y;
}

void Crasher::crash()
{
    y = 0;    // this crashes the game
}
```

### PR DESCRIPTION
**Pull request by melex750**
Fixes #207, #1016, #854, #882

Some saved games that activated those bugs are incompatible.

Fixed using out-of-class methods.

I think it's safe to close these too: #874 and #971

### CODE REVIEW

**Comment by melex750** [comment]
What the ?

**Comment by krzys-h** [comment]
That's a lot of bugs fixed in one PR, awesome!

File: src/CBot/CBotClass.h [hunk 1]
```diff
@@ -303,6 +303,12 @@ class CBotClass
      */
     static void DefineClasses(std::list<CBotClass*> pClassList, CBotCStack* pStack);
 
+    /*!
+     * \brief Get the list of user-defined methods in this class.
+     * \return List of methods, can be empty.
+     */
+    const std::list<CBotFunction*>& GetFunctions();
```
Comment (CookiePLMonster) [inline src/CBot/CBotClass.h]:
```suggestion
    const std::list<CBotFunction*>& GetFunctions() const;
```

File: src/CBot/CBotInstr/CBotFunction.cpp [hunk 2]
```diff
@@ -76,6 +76,18 @@ bool CBotFunction::IsPublic()
     return m_bPublic;
 }
 
+////////////////////////////////////////////////////////////////////////////////
+bool CBotFunction::IsProtected()
```
Comment (CookiePLMonster) [inline src/CBot/CBotInstr/CBotFunction.cpp]:
```suggestion
bool CBotFunction::IsProtected() const
```

File: src/CBot/CBotInstr/CBotFunction.cpp [hunk 3]
```diff
@@ -76,6 +76,18 @@ bool CBotFunction::IsPublic()
...
+{
+    return m_bProtect;
+}
+
+////////////////////////////////////////////////////////////////////////////////
+bool CBotFunction::IsPrivate()
```
Comment (CookiePLMonster) [inline src/CBot/CBotInstr/CBotFunction.cpp]:
```suggestion
bool CBotFunction::IsPrivate() const
```

File: src/CBot/CBotInstr/CBotFunction.cpp [hunk 4]
```diff
@@ -1020,6 +1205,12 @@ std::string CBotFunction::GetParams()
     return params;
 }
 
+////////////////////////////////////////////////////////////////////////////////
+const std::string& CBotFunction::GetClassName()
```
Comment (CookiePLMonster) [inline src/CBot/CBotInstr/CBotFunction.cpp]:
```suggestion
const std::string& CBotFunction::GetClassName() const
```

File: src/CBot/CBotInstr/CBotFunction.h [hunk 5]
```diff
@@ -215,12 +271,30 @@ class CBotFunction : public CBotInstr
      */
     std::string GetParams();
 
+    /*!
+     * \brief Get the name of the class for a method.
+     * \return The name of a class or empty string if it's not a method.
+     */
+    const std::string& GetClassName();
```
Comment (CookiePLMonster) [inline src/CBot/CBotInstr/CBotFunction.h]:
```suggestion
    const std::string& GetClassName() const;
```

File: src/CBot/CBotInstr/CBotFunction.h [hunk 6]
```diff
@@ -215,12 +271,30 @@ class CBotFunction : public CBotInstr
...
      * \brief IsPublic
      * \return
      */
     bool IsPublic();
 
+    /*!
+     * \brief Check if a method is protected.
+     * \return true if a method was compiled with "protected" keyword.
+     */
+    bool IsProtected();
```
Comment (CookiePLMonster) [inline src/CBot/CBotInstr/CBotFunction.h]:
```suggestion
    bool IsProtected() const;
```

File: src/CBot/CBotInstr/CBotFunction.h [hunk 7]
```diff
@@ -215,12 +271,30 @@ class CBotFunction : public CBotInstr
...
+
+    /*!
+     * \brief Check if a method is private.
+     * \return true if a method was compiled with "private" keyword.
+     */
+    bool IsPrivate();
```
Comment (CookiePLMonster) [inline src/CBot/CBotInstr/CBotFunction.h]:
```suggestion
    bool IsPrivate() const;
```

File: src/CBot/CBotClass.h (hunk 1 above)
Comment (melex750) [reply src/CBot/CBotClass.h]: This new 'suggest changes' feature is nice, I didn't even know it existed.

**Comment by MrSimbax** [comment]
Could you merge dev so that the checks become green, and make the suggested `const` keyword changes?

**Comment by melex750** [comment]
Will do the merge and suggested changes in a few minutes...

**Comment by melex750** [comment]
Done.

**Comment by MrSimbax** [comment]
Jenkins finally built this, and unfortunately there is a compile error now.

```cpp
[ 69%] Building CXX object src/CBot/CMakeFiles/CBot.dir/CBotClass.cpp.obj

/var/lib/jenkins/workspace/colobot_colobot_PR-1047/src/CBot/CBotClass.cpp:545:33: error: prototype for 'const std::__cxx11::list<CBot::CBotFunction*>& CBot::CBotClass::GetFunctions()' does not match any in class 'CBot::CBotClass'

 const std::list<CBotFunction*>& CBotClass::GetFunctions()

                                 ^

In file included from /var/lib/jenkins/workspace/colobot_colobot_PR-1047/src/CBot/CBotClass.cpp:20:0:

/var/lib/jenkins/workspace/colobot_colobot_PR-1047/src/CBot/../CBot/CBotClass.h:310:37: error: candidate is: const std::__cxx11::list<CBot::CBotFunction*>& CBot::CBotClass::GetFunctions() const

     const std::list<CBotFunction*>& GetFunctions() const;

                                     ^

make[2]: *** [src/CBot/CMakeFiles/CBot.dir/build.make:89: src/CBot/CMakeFiles/CBot.dir/CBotClass.cpp.obj] Error 1

make[1]: *** [CMakeFiles/Makefile2:277: src/CBot/CMakeFiles/CBot.dir/all] Error 2

make: *** [Makefile:152: all] Error 2

script returned exit code 2
```

**Comment by melex750** [comment]
Fixed.

**Pull request merged**